### PR TITLE
set 'rstudio.notebook.executing' option during notebook exec

### LINKED ISSUE #3894: Set an option when running a notebook chunk
Some functions (such as progress bars) print a lot of output to the console, which can be prohibitively expensive when running in notebook.

Therefore it would be useful if the IDE could set a global option when code was being executed in a notebook chunk, so this output could be scaled back / turned off.

https://github.com/tidyverse/readr/issues/793 is a issue related to this

### PR DESCRIPTION
**Pull request by kevinushey**
Closes #3894.

### CODE REVIEW

File: src/cpp/session/modules/NotebookHtmlWidgets.R [hunk 1]
```diff
@@ -300,6 +303,9 @@
 
 .rs.addFunction("releaseHtmlCapture", function()
 {
+   # note that we're done executing
+   options(rstudio.notebook.executing = TRUE)
```
Comment (jmcphers) [inline src/cpp/session/modules/NotebookHtmlWidgets.R]: I think this is meant to be FALSE. :-)

File: src/cpp/session/modules/NotebookHtmlWidgets.R [hunk 2]
```diff
@@ -284,6 +284,9 @@
                                             libraryFolder,
                                             chunkOptions)
 {
+   # let others know we're executing
+   options(rstudio.notebook.executing = TRUE)
```
Comment (jmcphers) [inline src/cpp/session/modules/NotebookHtmlWidgets.R]: Is there a particular reason this is coupled to HTML widget capture? If not, how would you feel about moving it to a more generic home? cf. how we set/unset the 'width' option: 

https://github.com/rstudio/rstudio/blob/2334aeb79e0306107fe8561ed7ea173ce299d81f/src/cpp/session/modules/rmarkdown/NotebookExec.cpp#L222
...
https://github.com/rstudio/rstudio/blob/2334aeb79e0306107fe8561ed7ea173ce299d81f/src/cpp/session/modules/rmarkdown/NotebookExec.cpp#L450

Comment (kevinushey) [reply src/cpp/session/modules/NotebookHtmlWidgets.R]: That seems like a more natural place -- I'll make that change.

File: src/cpp/session/modules/NotebookHtmlWidgets.R (hunk 1 above)
Comment (kevinushey) [reply src/cpp/session/modules/NotebookHtmlWidgets.R]: :doh:

File: src/cpp/session/modules/rmarkdown/NotebookExec.cpp [hunk 3]
```diff
@@ -446,6 +451,9 @@ void ChunkExecContext::disconnect()
    if (error)
       LOG_ERROR(error);
 
+   // broadcast that we're done with notebook execution
+   r::options::setOption(kRStudioNotebookExecuting, false);
```
[the remark this answers is not in the document]
Comment (kevinushey) [reply src/cpp/session/modules/rmarkdown/NotebookExec.cpp]: Perhaps we can just set the option to `FALSE` on startup? Having that type stability is nice.

**Comment by kevinushey** [comment]
Good point. Ultimately users who want to check the option should use something like:

``` r
if (identical(getOption("rstudio.notebook.executing", FALSE), TRUE)) { ... }
```

to check the option so the unset option is handled as `FALSE`.

**Comment by jimhester** [comment]
Thanks for adding this!

**Comment by jimhester** [comment]
Also you can always use `isTRUE(getOption("rstudio.notebook.executing"))` as a user, if the option is unset `getOption()` returns `NULL`, so it always returns `FALSE` unless the option is actually set to `TRUE`.

**Pull request merged**